File: THCS/2022_2023/HaNoi_2223/Bai3/TPS.py
def check(x, a, n, k):
    t = a[0] + x
    d = 1  # Số trạm phát sóng đã sử dụng

    for i in range(1, n):
        if a[i] > t + x:  # Nếu trạm thu sóng thứ i không được bao phủ
            d += 1
            t = a[i] + x  # Đặt trạm phát sóng tại vị trí a[i]

    t = a[1] + x  # Khởi tạo lại t cho lần kiểm tra thứ hai
    d1 = 1  # Số trạm phát sóng thứ hai

    for i in range(2, n + 1):
        if a[i] > t + x:  # Tương tự như trên
            d1 += 1
            t = a[i] + x  # Đặt trạm phát sóng tại vị trí a[i]

    return min(d, d1) <= k  # Trả về true nếu số trạm phát sóng không vượt quá K

def binary_search(l, r, a, n, k):
    res = 0  # Biến để lưu kết quả
    while l <= r:
        mid = (l + r) // 2
        if check(mid, a, n, k):
            res = mid  # Cập nhật kết quả
            r = mid - 1  # Tìm kiếm phạm vi nhỏ hơn
        else:
            l = mid + 1  # Tìm kiếm phạm vi lớn hơn
    return res  # Trả về phạm vi phát sóng ngắn nhất

def main():
    # Đọc dữ liệu từ tệp
    with open("TPS.INP", "r") as inp:
        n = int(inp.readline().strip())  # Số trạm thu sóng
        a = [int(inp.readline().strip()) for _ in range(n)]  # Đọc vị trí trạm thu sóng
        k = int(inp.readline().strip())  # Số trạm phát sóng

    a.sort()  # Sắp xếp vị trí các trạm thu sóng
    a.append(1000000 + a[0])  # Thêm trạm thu sóng đầu tiên vào cuối để xử lý vòng tròn

    # Tìm và in ra phạm vi phát sóng ngắn nhất
    result = binary_search(1, 1000000, a, n, k)
    with open("TPS.OUT", "w") as out:
        out.write(str(result) + "\n")

if __name__ == "__main__":
    main()

File: THCS/2022_2023/HaNoi_2223/Bai3/TPS.cpp
#include<bits/stdc++.h>
using namespace std;
#define ll long long

int a[1005], n, k;

// Hàm kiểm tra xem có thể phủ sóng tất cả trạm thu sóng với phạm vi R hay không
bool check(ll x) {
    ll t = a[1] + x, d = 1;  // d là số trạm phát sóng đã sử dụng
    for(int i = 2; i <= n; i++) {
        if(a[i] > t + x) {  // Nếu trạm thu sóng thứ i không được bao phủ
            d++;
            t = a[i] + x;  // Đặt trạm phát sóng tại vị trí a[i]
        }
    }
    t = a[2] + x;  // Khởi tạo lại t cho lần kiểm tra thứ hai
    ll d1 = 1;  // Số trạm phát sóng thứ hai
    for(int i = 3; i <= n + 1; i++) {
        if(a[i] > t + x) {  // Tương tự như trên
            d1++;
            t = a[i] + x;  // Đặt trạm phát sóng tại vị trí a[i]
        }
    }
    return min(d, d1) <= k;  // Trả về true nếu số trạm phát sóng không vượt quá K
}

// Hàm tìm kiếm nhị phân để tìm phạm vi phát sóng ngắn nhất
ll bs(ll l, ll r) {
    ll res = 0;  // Biến để lưu kết quả
    while(l <= r) {
        ll mid = (l + r) / 2;
        if(check(mid)) {
            res = mid;  // Cập nhật kết quả
            r = mid - 1;  // Tìm kiếm phạm vi nhỏ hơn
        } else {
            l = mid + 1;  // Tìm kiếm phạm vi lớn hơn
        }
    }
    return res;  // Trả về phạm vi phát sóng ngắn nhất
}

int main() {
    // Đọc dữ liệu từ tệp
    freopen("TPS.INP", "r", stdin);
    freopen("TPS.OUT", "w", stdout);
    
    cin >> n;  // Số trạm thu sóng
    for(int i = 1; i <= n; i++) {
        cin >> a[i];  // Đọc vị trí trạm thu sóng
    }
    cin >> k;  // Số trạm phát sóng
    
    sort(a + 1, a + n + 1);  // Sắp xếp vị trí các trạm thu sóng
    a[n + 1] = 1000000 + a[1];  // Thêm trạm thu sóng đầu tiên vào cuối để xử lý vòng tròn
    
    cout << bs(1, 1000000) << endl;  // Tìm và in ra phạm vi phát sóng ngắn nhất
    return 0;  // Kết thúc chương trình
}
